### DISCUSSION TITLE
Unable to make connection to Argus 3 Pro

### DISCUSSION THREAD

**Discussion by mclardass** [thirtythreeforty/neolink#336] [Q&A] [answered]
I saw in other posts that users were able to connect to these cameras using neolink 18.04. Since that version is no longer available I've been trying 20.04 but getting the following errors:

./neolink rtsp -c config.toml 
[2023-01-23T04:13:36Z INFO  neolink] Neolink 7087b348c40929c8e077880bdaca5bab88cd1eab release
[2023-01-23T04:13:36Z ERROR neolink::rtsp] Error streaming from camera backyard, will retry in 1s: Failed to connect to camera backyard at UID: XXXXXXXXXXXXX on channel 0
    
    Caused by:
        0: Connection error
        1: Communication error
        2: Operation not permitted (os error 1)

./neolink pir backyard off -c config.toml 
[2023-01-23T04:16:17Z INFO  neolink] Neolink 7087b348c40929c8e077880bdaca5bab88cd1eab release
[2023-01-23T04:16:17Z INFO  neolink::utils] backyard: Connecting to camera at UID: XXXXXXXXXXXXXX
Error: Failed to connect to camera backyard at UID: XXXXXXXXXXXXXXX on channel 0

Config looks like this:
[[cameras]]
name = "backyard"
username = "admin"
password = "XXXXXXXXXXXX"
uid = "XXXXXXXXXXXXXXX"
stream = "mainStream"   ##tried substream
encryption = 2   ## (tried the 3 options)

Not seeing anything in extended debug:
0:00:00.717252997  1650  0x1487600 DEBUG             rtspserver rtsp-server.c:891:gst_rtsp_server_create_socket:<GstRTSPServer@0x148fdd0> opened sending server socket
0:00:00.717400339  1650  0x1487600 DEBUG             rtspserver rtsp-server.c:919:gst_rtsp_server_create_socket:<GstRTSPServer@0x148fdd0> listening on server socket 0x15850f8 with queue of 5
[2023-01-23T04:25:26Z ERROR neolink::rtsp] Error streaming from camera backyard, will retry in 1s: Failed to connect to camera backyard at UID: 95270003UY921IBY on channel 0

Flat network without internal firewalls, can ping the cameras from the neolink box.

**Comment by QuantumEntangledAndy** [comment, marked as answer]
Neolink's version is only up to 0.4 in this fork. I presume you mean gstreamer? Anyways `Operation not permitted (os error 1)` means that you do not have permission to open up a port on your machine. Is this a linux box/docker? Given the error I think this is happening during the attempt to do a UID discovery and negotiation. Can you set `export RUST_LOG=debug` which should give more information about the  discovery process.

**Reply by mclardass** [reply]
Sorry, that was the Ubuntu version on the neolink download. Using a Linux box with iptables and I set ACCEPT all for the camera IPs but that didn't work. Flushing the rules allowed me to establish a connection so clearly I'm blocking a required port. The only thing additional in the debug output was:

[2023-01-24T00:18:17Z DEBUG neolink_core::bc_protocol::connection::udpconn::discover] Broadcasting to: [(255.255.255.255, 2015), (255.255.255.255, 2018), (192.168.1.255, 2015), (192.168.1.255, 2018)]

Then permitting tcp/udp ports 2015 and 2018 gives this:
[2023-01-24T00:23:30Z INFO  neolink_core::bc_protocol::connection::udpconn::discover] Trying remote discovery against reolink servers
[2023-01-24T00:23:30Z DEBUG neolink_core::bc_protocol::connection::udpconn::discover] Trying register: p2p.reolink.com

Added a rule for p2p.reolink.com but still not working but clearly this is on my side. Appreciate the help!

**Reply by QuantumEntangledAndy** [reply]
If I had to guess you've got something blocking broadcasts or the camera is on a different subnet.

**Reply by QuantumEntangledAndy** [reply]
The broadcast may also be getting blocked by your router/switch.

**Reply by QuantumEntangledAndy** [reply]
Do you have anything in your ip tables about `--pkt-type multicast`

**Reply by QuantumEntangledAndy** [reply]
The connection to p2p.reolink is actually a fallback. We prefer broadcast based local discovery but when that fails we try to connect to the p2p servers to learn it that way.

**Reply by QuantumEntangledAndy** [reply]
Also full list of p2p servers we use 

"p2p.reolink.com",
        "p2p1.reolink.com",
        "p2p2.reolink.com",
        "p2p3.reolink.com",
        "p2p6.reolink.com",
        "p2p7.reolink.com",
        "p2p8.reolink.com",
        "p2p9.reolink.com",
        "p2p14.reolink.com",
        "p2p15.reolink.com",

**Reply by mclardass** [reply]
Got sidetracked by work but obviously your detailed responses will put me on the right path. I don't see any multicast rules in my ip tables and will likely have to go through them section by section when I have time. Appreciate the efforts in supporting this camera family, couldn't run POE to my required areas and thought I wouldn't be able to integrate the three Argus 3 with HA. Glad you've provided a possible solution, now to make it work in my setup. 
Cheers!

**Reply by QuantumEntangledAndy** [reply]
One thing to consider trying is to clear your ip rules again. Then connect with `RUST_LOG=debug` to see from the log if it is connecting via local discovery or via the `p2p.reolink`. If it is still connecting via the p2p servers even with your rules clear then it is not the local machine that is stopping the broadcasts. You should then check your router to see if it is stopping the broadcasts.

**Comment by mclardass** [comment]
Finally got a chance to test, if I remove these three iptable rules I can successfully connect:

-A INPUT -j REJECT --reject-with icmp-port-unreachable
-A FORWARD -j REJECT --reject-with icmp-port-unreachable
-A OUTPUT -j REJECT --reject-with icmp-port-unreachable

I used VLC to point to the neolink host (port 8554/camera_name) and the camera stream came right up. Next step, go back and figure out how to set up the integration in HA. Thanks again for all of your help!

EDIT: Added generic camera integration and picture entity cards and have all three cameras on my HA dashboard! Realizing too late that neolink on an old pi will not decode very well, time to move to a beefier box.
